src/emotes/graphics_protocol.rs

```diff
@@ -0,0 +1,358 @@
+
+        match image.format() {
+            None => Err(anyhow!("Could not guess image format.")),
+            Some(ImageFormat::WebP) => Err(anyhow!("WebP image format is not supported.")),
+            Some(ImageFormat::Gif) => Ok(Self::Gif(Gif::new(id, image)?)),
+            Some(_) => Ok(Self::Static(Static::new(id, image)?)),
+        }
+    }
+}
+
+impl Command for Load {
+    fn write_ansi(&self, f: &mut impl fmt::Write) -> fmt::Result {
+        match self {
+            Self::Static(s) => s.write_ansi(f),
+            Self::Gif(g) => g.write_ansi(f),
+        }
+    }
+
+    #[cfg(windows)]
+    fn execute_winapi(&self) -> std::result::Result<(), std::io::Error> {
+        panic!("Windows version not supported.")
+    }
+}
+
+impl Size for Load {
+    fn size(&self) -> (u32, u32) {
+        match self {
+            Self::Static(s) => s.size(),
+            Self::Gif(g) => g.size(),
+        }
+    }
+}
+
+pub struct Display {
+    id: u32,
+    pid: u32,
+    width: u32,
+    offset: u32,
+    x: u16,
+    y: u16,
+}
+
+impl Display {
+    pub const fn new((x, y): (u16, u16), id: u32, pid: u32, width: u32, offset: u32) -> Self {
+        Self {
+            id,
+            pid,
+            width,
+            offset,
+            x,
+            y,
+        }
+    }
+}
+
+impl Command for Display {
+    fn write_ansi(&self, f: &mut impl fmt::Write) -> fmt::Result {
+        MoveTo(self.x, self.y).write_ansi(f)?;
+        // r=1: Set height to 1 row
+        write!(
+            f,
+            gp!("a=p,i={id},p={pid},r=1,c={width},X={offset},q=2;"),
+            id = self.id,
+            pid = self.pid,
+            width = self.width,
+            offset = self.offset
+        )
+    }
+
+    #[cfg(windows)]
+    fn execute_winapi(&self) -> std::result::Result<(), std::io::Error> {
+        panic!("Windows version not supported.")
+    }
+}
+
+#[derive(Eq, PartialEq)]
+pub struct Clear(pub u32, pub u32);
+
+impl Command for Clear {
+    fn write_ansi(&self, f: &mut impl fmt::Write) -> fmt::Result {
+        if *self == Self(0, 0) {
+            // Delete and unload all images
+            write!(f, gp!("a=d,d=A,q=2;"))
+        } else if self.0 == 0 {
+            // Delete all images
+            write!(f, gp!("a=d,d=a,q=2;"))
+        } else {
+            write!(
+                f,
+                gp!("a=d,d=i,i={id},p={pid},q=2;"),
+                id = self.0,
+                pid = self.1,
+            )
+        }
+    }
+
+    #[cfg(windows)]
+    fn execute_winapi(&self) -> std::result::Result<(), std::io::Error> {
+        panic!("Windows version not supported.")
+    }
+}
+
+fn query_terminal(command: &[u8]) -> Result<String> {
+    let c = *command.last().context("Command is empty")? as char;
+    let mut stdout = Term::stdout();
+    stdout.write_all(command)?;
+    stdout.flush()?;
+
+    loop {
+        let c = stdout.read_key()?;
+        if let Key::UnknownEscSeq(_) = c {
+            break;
+        }
+    }
```

Comment:
Can this function be documented/commented? For example, I'm not sure the reason for this loop, and whether or not there is a potential for an infinite loop. Would be nice to have an explanation for the response building code used below, or a link to the reference used.

Reply:
This can cause an infinite loop on terminals that do not respond to the <ESC>]14t/c queries, but it is only called if the terminal is kitty or wezterm, which do respond to theses queries. I read until I find an escape sequence because if the user start the program an inputs something, there might be some chars we need to read before we can read the réponse from the terminal. This is also why this function is called before we start the tui. 
If for some reason kitty/wezterm do not respond to the queries, it'll be an infinite loop. But I haven't found any other way. Maybe a readkey with timeout function exist?